### ISSUE TITLE
`prefer-const` rule warns when "let" is needed for scoping

### ISSUE THREAD

**Issue by ljharb** [eslint/eslint#5822]
on eslint v2.7.0.

`echo 'let foo; const bar = function bar() { return foo; }; foo = 3; console.log(bar());' | ./node_modules/.bin/eslint --stdin --rule 'prefer-const: 2' --no-eslintrc --parser-options=ecmaVersion:6`
gives
```sh
<text>
  1:54  error  'foo' is never reassigned, use 'const' instead  prefer-const

✖ 1 problem (1 error, 0 warnings)
```

I'd like it not to warn here, since the `let` is required to be able to refer to it without violating the `no-use-before-define` rule. If it does warn, then must I disable one of the rules since they are in conflict here?

**Comment by platinumazure** [comment]
Why not do `const foo = 3;` at the beginning?

Or if, as I suspect, this is a simplified use case, could you please post your actual use case?

**Comment by ljharb** [comment]
This is indeed simplified.

The real use case is recursive React PropType shapes:
```js
let ItemShape;

// Allows for recursive type checking
const lazyItemShape = function ItemShape(...args) {
  return ItemShape.apply(this, args);
};

ItemShape = PropTypes.shape({
  items: PropTypes.arrayOf(lazyItemShape),
});
```

**Comment by nzakas** [comment]
I don't think there's an easy way to do this. Your best bet is to just manually disable the rule on that line. 

Also, please stick to using the issue template when reporting bugs. It helps us a lot.

**Comment by lukeapage** [comment]
this is a duplicate (or at least extremely similiar) to #5692. In there the suggestion was you would have to write

```
const ItemShape = PropTypes.shape({
  items: PropTypes.arrayOf(lazyItemShape),
});

// Allows for recursive type checking
function lazyItemShape(...args) {
  return ItemShape.apply(this, args);
};
```

and then turn off the rule for no-use-before-define.

as in that issue, an alternative to these is to write this (I assume it will also work here):
```
let ItemShape = null;

// Allows for recursive type checking
const lazyItemShape = function ItemShape(...args) {
  return ItemShape.apply(this, args);
};

ItemShape = PropTypes.shape({
  items: PropTypes.arrayOf(lazyItemShape),
});
```

@mysticatea suggested on that issue an option in prefer-const to not warn if the reference was read before assignment and I think this would resolve both this bug and #5692

**Comment by nzakas** [comment]
@mysticatea thoughts?

**Comment by ljharb** [comment]
@nzakas sorry, i'll try to stick to the template in the future.

**Comment by mysticatea** [comment]
@lukeapage Thank you for the explanation!

As I mentioned at https://github.com/eslint/eslint/issues/5692#issuecomment-203690886, I think I can add an option to ignore variables that are read before its first assignment.

**Comment by nzakas** [comment]
@mysticatea thanks. So is this a duplicate of #5692?

**Comment by mysticatea** [comment]
@nzakas I think so.

**Comment by ljharb** [comment]
Closing as a duplicate of #5692, then :-)
